File: Baekjoon/11000~11999/[11054]가장 긴 바이토닉 부분 수열/solution.js
const fs = require("fs");
// const inputData = fs.readFileSync("example.dat").toString().split("\n");
const inputData = fs.readFileSync("/dev/stdin").toString().split("\n");
const N = Number(inputData[0]);
const seq = inputData[1].split(" ").map((item) => Number(item));
const rseq = [...seq].reverse();
let dp_asc = Array(N).fill(1);
let dp_desc = Array(N).fill(1);

for (let i = 1; i < N; i++) {
  for (let j = i - 1; j >= 0; j--) {
    if (seq[i] > seq[j]) {
      dp_asc[i] = Math.max(dp_asc[i], dp_asc[j] + 1);
    }
    if (rseq[i] > rseq[j]) {
      dp_desc[i] = Math.max(dp_desc[i], dp_desc[j] + 1);
    }
  }
}

let lbs_size = 0;
for (let i = 0; i < N; i++) {
  lbs_size = Math.max(lbs_size, dp_asc[i] + dp_desc[N - i - 1] - 1);
}

console.log(lbs_size);

File: Baekjoon/11000~11999/[11054]가장 긴 바이토닉 부분 수열/solution.py
import sys

N = int(sys.stdin.readline())
seq = list(map(int, sys.stdin.readline().split()))
rseq = list(reversed(seq))
dp_asc = [1 for i in range(N)]
dp_desc = [1 for i in range(N)]

for i in range(1, N):
    for j in range(i - 1, -1, -1):
        if seq[i] > seq[j]:
            dp_asc[i] = max(dp_asc[i], dp_asc[j] + 1)
        if rseq[i] > rseq[j]:
            dp_desc[i] = max(dp_desc[i], dp_desc[j] + 1)

lbs_size = 0
for i in range(N):
    lbs_size = max(lbs_size, dp_asc[i] + dp_desc[N - i - 1] - 1)

sys.stdout.write(str(lbs_size) + "\n")
